src/System.Private.CoreLib/src/System/Threading/ManagedThreadId.cs

```diff
@@ -9,15 +9,171 @@
-using System.Runtime.InteropServices;
-using System.Security;
 
 namespace System.Threading
 {
     internal class ManagedThreadId
     {
+        //
+        // Binary tree used to keep track of active thread ids. Each node of the tree keeps track of 32 consecutive ids.
+        // Implemented as immutable collection to avoid locks. Each modification creates a new top level node.
+        // 
+        class ImmutableIdDispenser
+        {
+            private readonly ImmutableIdDispenser _left; // Child nodes
+            private readonly ImmutableIdDispenser _right;
+
+            private readonly int _used; // Number of ids tracked by this node and all its childs
+            private readonly int _size; // Maximum number of ids that can be tracked by this node and all its childs
+
+            private readonly uint _bitmap; // Bitmap of ids tracked by this node
+
+            private const int BitsPerNode = 32;
+
+            private ImmutableIdDispenser(ImmutableIdDispenser left, ImmutableIdDispenser right, int used, int size, uint bitmap)
+            {
+                _left = left;
+                _right = right;
+                _used = used;
+                _size = size;
+                _bitmap = bitmap;
+
+                CheckInvariants();
+            }
+
+            [Conditional("DEBUG")]
+            private void CheckInvariants()
+            {
+                int actualUsed = 0;
+
+                uint countBits = _bitmap;
+                while (countBits != 0)
+                {
+                    actualUsed += (int)(countBits & 1);
+                    countBits >>= 1;
+                }
+
+                if (_left != null)
+                {
+                    Debug.Assert(_left._size == ChildSize);
+                    actualUsed += _left._used;
+                }
+                if (_right != null)
+                {
+                    Debug.Assert(_right._size == ChildSize);
+                    actualUsed += _right._used;
+                }
+
+                Debug.Assert(actualUsed == _used);
+                Debug.Assert(_used <= _size);
+            }
+
+            private int ChildSize
+            {
+                get
+                {
+                    return (_size / 2) - (BitsPerNode / 2);
+                }
+            }
+
+            public static ImmutableIdDispenser Empty
+            {
+                get
+                {
+                    // The empty dispenser has the id=0 allocated. It ensures that ManagedThreadIdNone
+                    // will not be ever given out.
+                    return new ImmutableIdDispenser(null, null, 1, BitsPerNode, 1);
+                }
+            }
+
+            public ImmutableIdDispenser AllocateId(out int id)
+            {
+                if (_used == _size)
+                {
+                    id = _size;
+                    return new ImmutableIdDispenser(this, null, _size + 1, checked(2 * _size + BitsPerNode), 1);
```

Comment:
The program would have to create billion threads and manage to starve the finalizer so that no thread ids are recycled for this to happen. It will never happen in practice. The checked statement is just "defense in depth".

In the super unlikely case that this can somehow happen, returning duplicate max id all the time would result in rapid program crash anyway because of locks would stop working. It is better to throw here where things went wrong and not hope for secondary effects.